### ISSUE TITLE
Report filters

### ISSUE THREAD

**Issue by mbouaziz** [aantron/bisect_ppx#305]
Say I want to ensure that a code diff achieves some coverage score. Adding annotations in the code would be cumbersome. Using exclusions would be still be pretty heavy on a large project when the diff touches only a few lines.

What I'd like is a way to specify ranges of source code I'm interested in, either at report time (coverage data can be reused for other ranges), or at compile time (more efficient if I don't need to reuse the coverage data).

For compile time, I suggest adding an option similar to `--exclusions` but that would do the opposite, e.g. `--only`.
I would also extend the exclude file format to support line ranges. E.g. (or `--` instead of `..`?)
```
file "baz.ml" [
  lines 4 .. 8
  line 12
]
```
which I could _easily_ extract from a `git diff`.

**Comment by aantron** [comment]
Agreed, Bisect_ppx needs some mechanisms for filtering the coverage data. I think for them to be expressive and fully usable, they will likely have to be report-time. See also https://github.com/aantron/bisect_ppx/issues/131.

The deciding factor for where this should be handled is less run-time efficiency, and more usability. I haven't fully thought it through yet, but I wonder if people would want to click through several related filters, or modify them gradually. If so, we definitely want to have all the data available in the reporter *and* the generated report, and have the UI display it with filters applied.

This would be less directly usable with the output of git diff, however.

**Comment by mbouaziz** [comment]
If I understand #131 correctly, it opens questions on :
- the two dimensions that I will badly call
  - _time_ (which test generates which coverage), mostly already solved by `BISECT_FILE`,
  - _space_ (what parts of the code we are interested in), which I want to cover in this issue;
- incrementality/differential of two coverages, which I'll come to in another issue a bit later ;)

I agree that filtering at compile-time is a bit overworked for what I need and doing it later works perfectly fine.
Doing it _only_ in the UI seems a bit too late though as one may want to get raw coverage score using `bisect-ppx-report summary`.

I imagine we could have `bisect-ppx-report html` take `--filter <filter-file>` arguments, each of them creating a tab in the UI. `bisect-ppx-report summary` would take at most one.

`<filter-file>` would be the name of a file, with the same syntax (to avoid too many syntaxes and parsers) as exclusions file (+ lines, as described above).
In my case I'm interested in inclusions (nothing except what is specified in the file) but I guess some may want exclusions (everything except what is specified in the file).
I suppose this could be solved by creating two arguments: `--include-filter` / `--exclude-filter`

**Comment by aantron** [comment]
> * _еime_ (which test generates which coverage), mostly already solved by `BISECT_FILE`,

How does `BISECT_FILE` address this, except at the very coarse level of running separate testing processes?



> * incrementality/differential of two coverages, which I'll come to in another issue a bit later ;)

Bisect in fact used to support this, but we removed it because there were no known users. See `-combine-expr` before https://github.com/aantron/bisect_ppx/issues/105.

<br/>

I gave this issue some more thought, too, and I got the impression that we don't have enough information (i.e. real use cases) to properly design the UI I was suggesting. So, I would be happy to merge any reasonable PR adding this feature in whatever form is convenient for you, with the understanding that we may modify or replace it with something equivalent later, if/when we get a better idea of who needs what else from it. In other words, it would have a sort of experimental status. This might be a bit annoying later if/when it changes, but Bisect_ppx is only a dev tool, so it shouldn't be painful.

How does that sound? I hope it can unblock getting this addressed efficiently :)

**Comment by mbouaziz** [comment]
> How does `BISECT_FILE` address this, except at the very coarse level of running separate testing processes?

I misunderstood the desired granularity. Indeed knowing which particular test stresses a given location/function/module/file would be quite interesting for contributors, I may come back to that later then ;)

I'll give a try to filters as I described, we'll see :)

**Comment by aantron** [comment]
I'm going to close this issue for now, as there is no activity and I'm not sure if this is still needed (we have worked on several other things since). Please do comment if you'd like to resume this later.
